### ISSUE TITLE
Compile Error with msys2

### ISSUE THREAD

**Issue by maosuyun2009** [MaskRay/ccls#327]
I try to compile ccls in win10 with msys2 but some error happened

gcc version: 7.3.0

```
D:\Applications\mingw64\bin\c++.exe  -DDEFAULT_RESOURCE_DIRECTORY=R\"(D:\Applications\mingw64\lib\clang\7.0.1)\" -I../src -isystem ../third_party -isystem D:/Applications/mingw64/include -isystem ../third_party/rapidjson/include -D__STDC_FORMAT_MACROS -O3 -DNDEBUG   -Wall -Wno-sign-compare -Wno-return-type -Wno-unused-result -fno-rtti -std=c++1z -MD -MT CMakeFiles/ccls.dir/src/pipeline.cc.obj -MF CMakeFiles\ccls.dir\src\pipeline.cc.obj.d -o CMakeFiles/ccls.dir/src/pipeline.cc.obj -c ../src/pipeline.cc
../src/pipeline.cc: In function 'std::__cxx11::string ccls::pipeline::{anonymous}::GetCachePath(std::__cxx11::string)':
../src/pipeline.cc:157:22: warning: unused variable '_' [-Wunused-variable]
   for (auto &[root, _] : g_config->workspaceFolders)
                      ^
../src/pipeline.cc: In function 'void ccls::pipeline::Standalone(const string&)':
../src/pipeline.cc:722:26: warning: unused variable '_' [-Wunused-variable]
     for (auto &[_, folder] : project.root2folder)
                          ^
In file included from D:/Applications/mingw64/include/c++/7.3.0/memory:64:0,
                 from ../src/utils.hh:22,
                 from ../src/serializer.hh:18,
                 from ../src/config.hh:18,
                 from ../src/lsp.hh:18,
                 from ../src/pipeline.hh:3,
                 from ../src/pipeline.cc:16:
D:/Applications/mingw64/include/c++/7.3.0/bits/stl_construct.h: In instantiation of 'void std::_Construct(_T1*, _Args&& ...) [with _T1 = std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >; _Args = {const std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >&}]':
D:/Applications/mingw64/include/c++/7.3.0/bits/stl_uninitialized.h:83:18:   required from 'static _ForwardIterator std::__uninitialized_copy<_TrivialValueTypes>::__uninit_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >*, std::vector<std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> > > >; _ForwardIterator = std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >*; bool _TrivialValueTypes = false]'
D:/Applications/mingw64/include/c++/7.3.0/bits/stl_uninitialized.h:134:15:   required from '_ForwardIterator std::uninitialized_copy(_InputIterator, _InputIterator, _ForwardIterator) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >*, std::vector<std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> > > >; _ForwardIterator = std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >*]'
D:/Applications/mingw64/include/c++/7.3.0/bits/stl_uninitialized.h:289:37:   required from '_ForwardIterator std::__uninitialized_copy_a(_InputIterator, _InputIterator, _ForwardIterator, std::allocator<_Tp>&) [with _InputIterator = __gnu_cxx::__normal_iterator<const std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >*, std::vector<std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> > > >; _ForwardIterator = std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >*; _Tp = std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >]'
D:/Applications/mingw64/include/c++/7.3.0/bits/stl_vector.h:331:31:   required from 'std::vector<_Tp, _Alloc>::vector(const std::vector<_Tp, _Alloc>&) [with _Tp = std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> >; _Alloc = std::allocator<std::pair<long long unsigned int, ccls::FuncDef<ccls::Vec> > >]'
../src/query.hh:104:8:   required from 'struct std::is_trivially_copy_constructible<ccls::IndexUpdate>'
D:/Applications/mingw64/include/c++/7.3.0/optional:105:8:   required from 'class std::_Optional_base<ccls::IndexUpdate>'
D:/Applications/mingw64/include/c++/7.3.0/optional:453:11:   required from 'class std::optional<ccls::IndexUpdate>'
../src/pipeline.cc:664:34:   required from here
D:/Applications/mingw64/include/c++/7.3.0/bits/stl_construct.h:75:7: error: use of deleted function 'std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = long long unsigned int; _T2 = ccls::FuncDef<ccls::Vec>]'
     { ::new(static_cast<void*>(__p)) _T1(std::forward<_Args>(__args)...); }
       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from D:/Applications/mingw64/include/c++/7.3.0/utility:70:0,
                 from D:/Applications/mingw64/include/c++/7.3.0/optional:36,
                 from ../src/utils.hh:18,
                 from ../src/serializer.hh:18,
                 from ../src/config.hh:18,
                 from ../src/lsp.hh:18,
                 from ../src/pipeline.hh:3,
                 from ../src/pipeline.cc:16:
D:/Applications/mingw64/include/c++/7.3.0/bits/stl_pair.h:292:17: note: 'std::pair<_T1, _T2>::pair(const std::pair<_T1, _T2>&) [with _T1 = long long unsigned int; _T2 = ccls::FuncDef<ccls::Vec>]' is implicitly deleted because the default definition would be ill-formed:
       constexpr pair(const pair&) = default;
                 ^~~~
D:/Applications/mingw64/include/c++/7.3.0/bits/stl_pair.h:292:17: error: use of deleted function 'ccls::FuncDef<ccls::Vec>::FuncDef(const ccls::FuncDef<ccls::Vec>&)'
In file included from ../src/query.hh:18:0,
                 from ../src/pipeline.hh:4,
                 from ../src/pipeline.cc:16:
../src/indexer.hh:163:8: note: 'ccls::FuncDef<ccls::Vec>::FuncDef(const ccls::FuncDef<ccls::Vec>&)' is implicitly deleted because the default definition would be ill-formed:
 struct FuncDef : NameMixin<FuncDef<V>> {
        ^~~~~~~
../src/indexer.hh:163:8: error: use of deleted function 'ccls::Vec<long long unsigned int>::Vec(const ccls::Vec<long long unsigned int>&)'
In file included from ../src/serializer.hh:18:0,
```

**Comment by MaskRay** [comment]
See https://github.com/MaskRay/ccls/pull/326 for another similar report. I think we should figure out the root cause first then work around that bug.

**Comment by MaskRay** [comment]
Can you apply this patch and see if msys2 compiles?

```patch
diff --git i/src/query.cc w/src/query.cc
index 6adfb424..5c07efed 100644
--- i/src/query.cc
+++ w/src/query.cc
@@ -105,3 +105,7 @@ QueryFunc::Def Convert(const IndexFunc::Def &o) {
   r.storage = o.storage;
+#if __clang__ || __GNUC__ >= 8
   return r;
+#else
+  return std::move(r);
+#endif
 }
@@ -125,3 +129,7 @@ QueryType::Def Convert(const IndexType::Def &o) {
   r.parent_kind = o.parent_kind;
+#if __clang__ || __GNUC__ >= 8
   return r;
+#else
+  return std::move(r);
+#endif
 }
```

**Comment by maosuyun2009** [comment]
[build_log.txt](https://github.com/MaskRay/ccls/files/2987774/build_log.txt)
can't compile success, same error

**Comment by MaskRay** [comment]
I missed one function. How about this?

```patch
diff --git i/src/query.cc w/src/query.cc
index 6adfb424..23d47bda 100644
--- i/src/query.cc
+++ w/src/query.cc
@@ -86,3 +86,7 @@ template <typename T> Vec<T> Convert(const std::vector<T> &o) {
   std::copy(o.begin(), o.end(), r.begin());
+#if __clang__ || __GNUC__ >= 8
   return r;
+#else
+  return std::move(r);
+#endif
 }
@@ -105,3 +109,7 @@ QueryFunc::Def Convert(const IndexFunc::Def &o) {
   r.storage = o.storage;
+#if __clang__ || __GNUC__ >= 8
   return r;
+#else
+  return std::move(r);
+#endif
 }
@@ -125,3 +133,7 @@ QueryType::Def Convert(const IndexType::Def &o) {
   r.parent_kind = o.parent_kind;
+#if __clang__ || __GNUC__ >= 8
   return r;
+#else
+  return std::move(r);
+#endif
 }
```

**Comment by maosuyun2009** [comment]
Same errror as before

**Comment by MaskRay** [comment]
Fixed. This was a bug that has been fixed in 7.4

I worked around it with

```cpp
template <typename T> struct Vec {
  std::unique_ptr<T[]> a;
  int s = 0;
#if !(__clang__ || __GNUC__ > 7 || __GNUC__ == 7 && __GNUC_MINOR__ >= 4)
  // Work around a bug in GCC<7.4 that optional<IndexUpdate> would not be
  // construtible.
  Vec() = default;
  Vec(const Vec &o) : a(std::make_unique<T[]>(o.s)), s(o.s) {
    std::copy(o.a.get(), o.a.get() + o.s, a.get());
  }
  Vec(Vec &&) = default;
  Vec &operator=(Vec &&) = default;
  Vec(std::unique_ptr<T[]> a, int s) : a(std::move(a)), s(s) {}
#endif
  const T *begin() const { return a.get(); }
  T *begin() { return a.get(); }
  const T *end() const { return a.get() + s; }
  T *end() { return a.get() + s; }
  int size() const { return s; }
  const T &operator[](size_t i) const { return a.get()[i]; }
  T &operator[](size_t i) { return a.get()[i]; }
};
```
